### DISCUSSION TITLE
Get Dynamically Updating Values while Python Script is Running

### DISCUSSION THREAD

**Discussion by jaroducsb** [bonsai-rx/bonsai#2025] [Q&A] [answered]
Hello,

I was wondering if the Python Scripting package had the capability of getting values that update dynamically while the script is still running. Currently, it seems that the Get and Eval modules only take the last value once the script finishes, and they only subscribe when the script is done. Ideally, I would like Get and/or Eval to output the current value of the variable, even if the program has not finished executing.

My goal with this is to create a user-friendly Python GUI for a Bonsai program we're running, which _might_ be too lofty of an idea. An example could be to change the trial duration in the GUI and have Bonsai get that value to set the duration of a Timer module. This change would happen while the GUI Is still open, so users could play around with parameters and test them in Bonsai. Currently, we can do this if we set all values in the GUI and then close it (which is amazing on its own), but it would be nice if it could remain open.

To make things easier, I was testing this with a simple Python script, and the ideal outputs would be 1, then 2, then 3, etc. until 100:

```
a = 0

def example():
    global a
    a += 1
    return a

while a < 100:
    example()
```

But currently, it would just output 100. Is there any way of doing this (ie. outputting 1, 2, 3, etc.) with the Scripting.Python modules? I have workarounds if not, it would just be extremely convenient to do it this way.

**Comment by ncguilbeault** [comment, marked as answer]
Hi there,

Yes, this is possible. The key lies in performing the example() function asynchronously. The current behavior you're seeing is due to python's Global Interpreter Lock (GIL), which ensures that the Python code you are running executes synchronously in a single thread. To access your variables dynamically, you'll need to run your function asynchronously in a separate thread. Fortunately, there are several options for doing this in Python.

I'm attaching an example using the Bonsai.Scripting.Python package to demonstrate one approach that I've had success with. It uses the python packages `asyncio` and `threading` to run the example function asynchronously in a seperate thread. The example function takes the global variable `a` and increments it every second. The function get_a() can then retrieve the global variable without being blocked.

In the bonsai workflow, you simply call the run function using `Eval` after loading your script which will spawn the thread and start the loop. Then, a `Timer` is used to retrieve the a variable every 2 seconds by running get_a inside of `Eval`. The `WriteLine` operator prints the value of a to the console.

Not sure if this is useful, but we use a similar approach for the [Bonsai.ML package](https://github.com/bonsai-rx/machinelearning/) to fit model parameters asynchronously during optimization while still allowing the model to perform online inference without being blocked. Hopefully this helps.

[async_python_example.zip](https://github.com/user-attachments/files/17148587/async_python_example.zip)

**Reply by jaroducsb** [reply]
Incredible, thank you so much! I'm excited to get started with the ML package; it's quickly becoming a current "need" rather than a future goal to get into.
